Dev/mstarzyk mobica/code style v2

mstarzyk-mobica opened the pull request:
## Todo list, based on discussion in #4783:
- [x] .clang-format file origin.
	Resolved by creating our own file.
	File layout is based on the file generated by `clang-format-12 --style=llvm --dump-config > .clang-format-12`.
	Removed options which do not apply to C code (such as Java code options or C++ templates)
- [x] parantheses after 'return'
- [x] parantheses after 'sizeof'
- [x] Enums definitions must stay with one definition per line 
- [x] //no-check-names comment should not be reformated, check-names.sh will not recognize them in other format.
- [x] Doxygen blocks
- [x] #define MBEDTLS_ERR_xxx need to remain in one line for scripts/generate_errors.pl parsing (it doesn't support continuation lines).

## Under discussion - please chime in if you feel like it:

### CMake files style.

Do we want to reformat existing CMake files? There is a separate tool available for it called `cmake-format`. A sample of it's output: https://github.com/ARMmbed/mbedtls/pull/4783/files?file-filters%5B%5D=.txt
On one hand this is yet another tool to use and maintain.
On the other hand, CMake files are 'code' after all and should follow certain style rules. Since we are already reforamtting all `.c` and `.h` files, we might as well adjust the `CMakeLists.txt` as well.

RC: Let's leave that for now.

### indentation: 4 or 8 spaces?

Seems reasonable value. Although K&R / Linux Kernel are using 8 spaces, it seems that current code would become less readable if the indentation was increased (certain parts of the library would have to be refactored by moving indented parts of the code to separate functions).

RC vote: Keep 4

### pointer alignment: right or left?

> @gilles-peskine-arm gilles-peskine-arm 5 days ago Member
> 
> Make that Right. It's what we currently use, its the most common style, and there's a good reason for it (the int* p, q; gotcha, even if we'd normally write int *p; int q; instead).

> @mstarzyk-mobica mstarzyk-mobica 2 days ago Author Member
> 
> What we are currently using is not a valid argument - current style is not really a good point of reference :P
> One could argue that both left and right alignments are valid - https://www.stroustrup.com/bs_faq2.html#whitespace
> Since we are not going for pure K&R style, I would prefer applying the pointer alignment which is emphasizing the type, rather then syntax, but this is strictly personal preference.

> @davidhorstmann-arm davidhorstmann-arm 2 days ago Member
> 
> My view probably doesn't carry much weight, but I agree with @gilles-peskine-arm on this.
> I can see the argument that the * should be a part of the type, because semantically that's what it is. However, that's more of an issue with the design of C - as C in practice links the * to the name, not the type.
> 
> The gotcha mentioned above is a problem, and even in cases where comma-separated declaration is used correctly, it looks untidy (int* p,* q?). If left-aligned pointer *s were to be used, I think it would also be best to ban comma-separated declarations, and make sure all declarations take place in separate statements.
> 
> Overall, I think the problems with left-aligned *s outweigh the semantic neatness, especially since anyone with more than a modicum of C experience is probably already used to dealing with the right-aligned *.

RC vote: keep right.

### column limit: 80 or 120?

> @gilles-peskine-arm gilles-peskine-arm 5 days ago Member
> 
> I'm ok with a hard limit at 120 and a soft limit at 80, but the result of the reflowing has lengthened a lot of lines beyond 80. That's unacceptable. Lines going beyond 80 columns > should be exceptional. The 80 character limit is not outdated: long lines are hard to read.

> @mstarzyk-mobica mstarzyk-mobica 2 days ago Author Member
> 
> I wouldn't call 120 chars a "long line". The main reason for 80 char limit was that old terminals could not fit more then 80 chars on the screen and handled line breaks badly. This is a good opportunity to move a bit "forward" and make use of the modern screens.

> @gilles-peskine-arm gilles-peskine-arm 2 days ago Member
> 
> I make use of modern screens by having multiple windows.
> 
> But more importantly, wider lines hurt readability, even with a wider window. For English text, 80 characters is already too wide, but code tends to have very short paragraphs and is indented, which puts 80 columns as a resaonable limit.
> 
> Please keep in mind that not everyone has top-notch eyesight and multiple ultra-wide monitors.

RC vote: hard limit at 120 (but for string literals probably) and a soft limit at 80 and we should try and document what are the exceptions in "Lines going beyond 80 columns > should be exceptional".

### clang-format version

Currently using clang-format-12. I picked the newest that was available in apt repo.
I believe that initial reformating of the code should be done with the most recent version of the tool.
Once the code is in desired style we will have to either:
- Upgrade clang-format on CI
- Use older clang-format (3.8) on CI with a "back-ported" config file. Although version 3.8 doesn't have as meany features as version 12, it might be enough to enforce code style on incoming commits

mstarzyk-mobica commented on include/mbedtls/ssl.h:
```diff
@@ -45,41 +45,63 @@
-#define MBEDTLS_ERR_SSL_INVALID_RECORD                    -0x7200  /**< An invalid SSL record was received. */
-#define MBEDTLS_ERR_SSL_CONN_EOF                          -0x7280  /**< The connection indicated an EOF. */
-#define MBEDTLS_ERR_SSL_DECODE_ERROR                      -0x7300  /**< A message could not be parsed due to a syntactic error. */
+#define MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS \
```
Even with penalty `PenaltyBreakComment: 0` these lines are split.
The solution here is to either modify scripts/generate_errors.pl to accept line continuation or add `clang-format-off/on` comments around `MBEDTLS_ERR_` defines.

ronald-cron-arm replied:
Another possibility is to rewrite them manually before to apply the reformatting tool. For example:
```C
/**< A cryptographic operation is in progress. Try again later. */
#define MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS                -0x7000
```

mstarzyk-mobica replied:
I added clang format guards. Please check if you're ok with it

ronald-cron-arm replied:
I would prefer to fix this manually rather than introducing clang format guards.

gilles-peskine-arm replied:
I'd prefer to lift the requirement of having the comment on the same line: it forces those lines to be very long. But that requires improvin `generate_errors.pl`.

mstarzyk-mobica replied:
I guess that this formatting
```
/**< A cryptographic operation is in progress. Try again later. */
#define MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS                -0x7000
```
will have to be applied manually.

mstarzyk-mobica replied:
Or maybe not manually.. https://regex101.com/r/Pyr0nT/1

gilles-peskine-arm replied:
Transforming the source code is a simple perl one-liner. Updating the consumer script is more than a one-liner.

TRodziewicz commented on include/mbedtls/base64.h:
```diff
@@ -26,8 +26,10 @@
-#define MBEDTLS_ERR_BASE64_BUFFER_TOO_SMALL               -0x002A  /**< Output buffer too small. */
-#define MBEDTLS_ERR_BASE64_INVALID_CHARACTER              -0x002C  /**< Invalid character in input. */
+#define MBEDTLS_ERR_BASE64_BUFFER_TOO_SMALL \
+    -0x002A /**< Output buffer too small. */
```
This is looks horrible. Is there any way in the tool options to define that the line break will happen in the comment rather than in the code? Or to move the comment above the code line?

mstarzyk-mobica replied:
see my comment above

TRodziewicz replied:
Ah, OK, didn't see.

mstarzyk-mobica replied:
Basically the clang-format is quite stubborn in this matter. I was not able to persuade to leave the assigned value as it was and adjust only the comment.

ronald-cron-arm replied:
As suggested for ssl.h, we could change them manually before to reformat automatically.

TRodziewicz commented on include/mbedtls/bignum.h:
```diff
@@ -966,18 +1006,22 @@ int mbedtls_mpi_inv_mod( mbedtls_mpi *X, const mbedtls_mpi *A,
-    MBEDTLS_MPI_GEN_PRIME_FLAG_DH =      0x0001, /**< (X-1)/2 is prime too */
-    MBEDTLS_MPI_GEN_PRIME_FLAG_LOW_ERR = 0x0002, /**< lower error rate from 2<sup>-80</sup> to 2<sup>-128</sup> */
+typedef enum
+{
```
I think that the Linux way is to keep the opening brace in the same line than `enum`.

mstarzyk-mobica replied:
hmm..
that would suggest that option "BreakBeforeBraces: Linux" is faulty. Thanks for noticing this

TRodziewicz commented on include/mbedtls/ecdsa.h:
```diff
@@ -53,13 +53,13 @@
-      /*V of r,s*/                ( ( bits ) + 8 ) / 8 ) )
+#define MBEDTLS_ECDSA_MAX_SIG_LEN(bits)               \
+    (/*T,L of SEQUENCE*/ ((bits) >= 61 * 8 ? 3 : 2) + \
+     /*T,L of r,s*/ 2 *                               \
```
I guess we need to reformat the original somehow to make the reformatted version readable.

mstarzyk-mobica replied:
There will be many such cases. Seems like my dream about using the clang-format as magic tool for "reformat and forget" is gone.
We will have to correct the code after clang-formating

ronald-cron-arm replied:
The important thing here is that when we are done with the reformatting, either some manual change then automatic reformatting or automatic reformatting then manual fixes, and we re-apply the automatic reformatting, nothing changes.

mstarzyk-mobica replied:
You are right, please see my comment https://github.com/ARMmbed/mbedtls/pull/4794#issuecomment-884835781

gilles-peskine-arm replied:
We want this to be usable by contributors on their own machine, not just by CI. It's important that a wide range of different versions of the tool give the same result.

mstarzyk-mobica replied:
I would rather pick one version and stick to it rather then attempt to support a variety of versions, for following reasons:

1. Not every option of clang-format is backwards compatible. It's also possible that consecutive versions of clang-format will yield different formatting from the same code.
2. Supporiting a wide range of versions would require using the `.clang-format` file which is using only options from the lowest version of supported range. This however gives no guarantee that code formatting will be consistent across various versions, because clang-format will use default values for undefined options. If we have `.clang-format` compatible with version 4 and format the code using clang-format-4 and clang-format-6, the results might differ solely due appearance of new style options in newer version. (Versions mentioned are just examples, I haven't checked specific differences between these two versions).

mstarzyk-mobica replied:
As agreed on video-call - we are going with version 12 only.

TRodziewicz reviewed (commented):
What about the *.function files?

mstarzyk-mobica commented:
> What about the *.function and *.fmt files?

I forgot about them :D

gilles-peskine-arm commented:
See also https://github.com/ARMmbed/mbedtls/pull/4799 for an exploration of what you can do with astyle (a lot less than clang-format, but at least no annoying line wraps).

mstarzyk-mobica commented:
I have some doubts about my initial statement in the PR description regarding clang version.
Clang formatting will be applied to the whole files in pushed commits. If clang-format-3.8 produces different outcome then clang-format-12, then perhaps we should not use clang-format-12 at all, and stick only to what is supported by CI.

mstarzyk-mobica commented:
Modified clang-format to conform with linux brace styles.
Fixed return and sizeof statements.
Introduced `clang-format off/on` guards around `MBEDTLS_ERR_*`

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -22,462 +22,451 @@
-#include "psa_crypto_se.h"
-#endif
-#include "psa_crypto_slot_management.h"
+#    if defined(MBEDTLS_PSA_CRYPTO_CONFIG)
```
I was not familiar with this but it seems quite good to spot at a glance the level of nesting of pre-processor guards. But this first indentation because of "#if defined(MBEDTLS_PSA_CRYPTO_C)" is a bit of a pity and does not bring much I would say. Would be there a way to avoid it ?

mstarzyk-mobica replied:
I'm not sure I understand.
Are you happy with preprocessor directives indentation or not?
If yes, are you suggesting that some of them should be indented and others not, regardless of nesting level?
Please clarify :)

ronald-cron-arm replied:
Having skimmed a bit more through the reformatted code preprocessor directives indentations make the code less readable to me. Thus I vote against but no veto, just my vote.

mstarzyk-mobica replied:
Alright, then it's:
* Against indented PP directives: 1 vote
* For indented PP directives: 1 vote (me :P)

Waiting for other opinions.
Also - there is also an option to indent with the '#', so it would be:
<code>&nbsp;&nbsp;&nbsp;&nbsp;#devine MACRO</code> instead of 
<code>#&nbsp;&nbsp;&nbsp;&nbsp;devine MACRO</code>

gilles-peskine-arm replied:
I have a weak preference for intended directives, with the `#` in column 1. I'd prefer them to have a shorter indentation than braced code: 2 instead of 4, or 1 instead of 2.

Like Ronald, I dislike having the whole-file conditional count as an indentation level. In a header file, which has the de facto standard `#ifndef HEADER_NAME_H` guard, that guard doesn't count:
```
#ifndef FOO_H
#define FOO_H
#if SOME_CONDITION
#    if NESTED_CONDITION
#        define WIBBLE
#    endif // NESTED_CONDITION
#endif // SOME_CONDITION
#endif // foo.h
```
But in our C files, which have a `MBEDTLS_MODULE_C` guard, that does count:
```
#if defined(MBEDTLS_FOO_C)
#    include "mbedlts/foo.h"
#    if CONDITION
#        define WOBBLE
#    endif // CONDITION
#endif // foo.c
```
Ronald's request was, can we have this instead in .c files?
```
#if defined(MBEDTLS_FOO_C)
#include "mbedlts/foo.h"
#if CONDITION
#    define WOBBLE
#endif // CONDITION
int f(void)
{
    return 42;
}
#endif // foo.c
```

And I'd prefer less indent on preprocessor directives:
```
#if defined(MBEDTLS_FOO_C)
#include "mbedlts/foo.h"
#if CONDITION
#  define WOBBLE
#endif // CONDITION
int f(void)
{
    return 42;
}
#endif // foo.c
```

mstarzyk-mobica replied:
I don't think it's possible to have some predefined level of nesting from which indentation would be applied. We could experiment with //clang-format on/off, but this would severly impact code readability.

Custom indent for preprocessor directovies comes with clang-format-13, the option is named `PPIndentWidth`.
clang-13 is not available in apt repo for ubuntu 18.04 (which I am currently using), so that's why I am using clang-12.
Further going into details on version differences should be preceded by agreeing on which clang-format version should the CI use.

mstarzyk-mobica commented:
Force pushed
Fixed script for parantheses modificiation
Fixed clang-format to recognize test cases wrappers (BEGIN_CASE, END_CASE)

TRodziewicz commented:
I've found very interesting and maybe even useful tool called `whatstyle`, which can identify the current code style and what's more "this program looks at your source code and generates a style definition that the reformatted source fits its original formatting as closely as possible." 
"The currently supported formatters are clang-format, YAPF, Tidy, indent, Artistic Style, Uncrustify, scalariform, scalafmt, rfmt and rustfmt."

It's here: https://github.com/mikr/whatstyle

mpg commented:
Labeling "do not merge" as this is a draft. This is mainly to that the list of unlabeled PRs remains empty. Please remove that label (and add others as appropriate) when this is no longer a draft.

mpg commented:
From the CI logs:
```
    89>..\..\programs\test\benchmark.c(235): error C2400: inline assembler syntax error in 'second operand'; found 'tsc' [C:\builds\workspace\mbed-tls-pr-head_PR-4794-head\worktrees\tmpyx0lfg99\visualc\VS2010\benchmark.vcxproj]
```
I _think_ that's the only error, it's just repeated a couple of time in various windows builds.
